### PR TITLE
recreate `ipfs-cache` indexeddb on `aborted` event

### PR DESCRIPTION
**Pull request by Tschuck**
Onboarding has stuck sometimes, when creating new profiles with `use-identity` flag enabled. IPFS cache kills itself by locking the whole indexeddb. Force recreating of the cache to clear the blocking of indexeddb.

- [CORE-1121]

[CORE-1121]: https://evannetwork.atlassian.net/browse/CORE-1121

### CODE REVIEW

File: src/app/ipfs-cache.ts [hunk 1]
```diff
@@ -27,38 +27,103 @@ let lastCache: IPFSCache;
...
-      openreq.onblocked = function (event) {
-        log('IndexDB blocked!');
+            break;
+          }
+          default: {
+            // do nothing
+          }
+        }
+      } else if (ex.number === -2147024882) {
+        // Internet Explorer 8
```
Comment (karladler) [inline src/app/ipfs-cache.ts]: 😅a good one 😂

File: src/app/ipfs-cache.ts [hunk 2]
```diff
@@ -27,38 +27,103 @@ let lastCache: IPFSCache;
...
-      openreq.onsuccess = () => {
-        try {
-          openreq.result.createObjectStore(storeName);
-        } catch (ex) { }
+  static isQuotaExceeded(ex: any): boolean {
+    let quotaExceeded = false;
 
-        resolve(openreq.result);
-      };
+    if (ex) {
```
Comment (karladler) [inline src/app/ipfs-cache.ts:46]: `if (ex.?code)` ?
I think we can remove the IE8 branch in the if

File: src/app/ipfs-cache.ts [hunk 3]
```diff
@@ -27,38 +27,103 @@ let lastCache: IPFSCache;
...
+    return this.db;
+  }
+
+  /**
+   * Close the current db, delete the previous one and create a new one.
+   */
+  private async recreateDb(): Promise<void> {
+    log('[ipfs-cache] recreate ipfs db - maybe aborted?');
+    // close previous db
+    await (await this.getOpenedDb()).close();
```
Comment (karladler) [inline src/app/ipfs-cache.ts:115]: shall we check beforehand if `this.db` exists, that we don't open a DB just to close it?

File: src/app/ipfs-cache.ts [hunk 4]
```diff
@@ -67,7 +132,7 @@ export class IPFSCache {
    * @param      {string}  hash    ipfs hash to store the data for
    * @param      {any}     data    data to store for the ipfs hash
    */
-  async add(hash: string, data: any) {
+  async add(hash: string, data: any): Promise<any> {
```
Comment (karladler) [inline src/app/ipfs-cache.ts]: I think set() has void return type.

```suggestion
  async add(hash: string, data: any): Promise<void> {
```

File: src/app/ipfs-cache.ts [hunk 5]
```diff
@@ -78,13 +143,35 @@ export class IPFSCache {
...
+    const transaction = db.transaction(this.storeName, type);
+    let request: any;
+
+    return new Promise((resolve, reject) => {
+      const handleError = (errorType: string): void => {
+        const isQuota = IPFSCache.isQuotaExceeded(transaction.error);
+        reject(new Error(`[ipfs-cache] - ${errorType} - ${transaction.error} - ${isQuota ? '(quota exceeded)' : ''}`));
+      };
+
+      transaction.oncomplete = (): void => resolve(request?.result);
```
Comment (karladler) [inline src/app/ipfs-cache.ts:163]: where is the `request` Object actually populated? Looks to me like we would always resolve to undefined ...

Comment (karladler) [reply src/app/ipfs-cache.ts:163]: nvm I just saw `request = actionCallback(transaction.objectStore(this.storeName));` at the end of this function.

The structure is really confusing, can we put the part from Line 173 inside the oncomplete function? Or even better first await transaction, make the code as linear as possible.

Comment (Tschuck) [reply src/app/ipfs-cache.ts:163]: We cannot move it into the oncomplete, because the oncomplete is the callback handle for the request.

**Review by karladler — APPROVED** [review]

File: src/systemjs/plugins/ens.ts [hunk 6]
```diff
@@ -90,7 +90,7 @@ export default function(System: any) {
         ensAddress.lastIndexOf('.' + rootDomain)
       ) + '.' + devDomain
 
-      // try to load from dev domain, if is 
+      // try to load from dev domain, if is
```
Comment (karladler) [inline src/systemjs/plugins/ens.ts]: there seems to be a part missing. If is "what"?

**Pull request merged**